plugins/kotlin/src/main/java/org/quiltmc/quiltflower/kotlin/pass/ResugarKotlinMethodsPass.java

```diff
@@ -0,0 +1,139 @@
+
+public class ResugarKotlinMethodsPass implements Pass {
+  @Override
+  public boolean run(PassContext ctx) {
+    boolean res = false;
+
+    DirectGraph digraph = FlattenStatementsHelper.build(ctx.getRoot());
+
+    for (DirectNode nd : digraph.nodes) {
+      List<Exprent> exprs = nd.exprents;
+      for (Exprent ex : exprs) {
+        res |= resugarExprs(ex);
+      }
+
+      for (int i = 0; i < exprs.size(); i++) {
+        Exprent expr = exprs.get(i);
+
+        ResugarRes exprRes = resugarExpr(expr);
+        if (exprRes.remove) {
+          exprs.remove(i);
+          i--;
+          res = true;
+        } else if (exprRes.expr != null) {
+          exprs.set(i, exprRes.expr);
+          res = true;
+        }
+      }
+    }
+
+    return res;
+  }
+
+  private static boolean resugarExprs(Exprent expr) {
+    boolean res = false;
+
+    for (Exprent ex : expr.getAllExprents()) {
+      res |= resugarExprs(ex);
+
+      Exprent map = resugarExpr(ex).expr;
+
+      if (map != null) {
+        expr.replaceExprent(ex, map);
+        res = true;
+      }
+    }
+
+    return res;
+  }
+
+
+  private static final MatchEngine[] NONNULL_INTRINSICS = {
+    new MatchEngine(
+      "exprent type:invocation invclass:kotlin/jvm/internal/Intrinsics signature:checkNotNullParameter(Ljava/lang/Object;Ljava/lang/String;)V"
+    ),
+    new MatchEngine(
+      "exprent type:invocation invclass:kotlin/jvm/internal/Intrinsics signature:checkNotNullExpressionValue(Ljava/lang/Object;Ljava/lang/String;)V"
+    ),
+    new MatchEngine(
+      "exprent type:invocation invclass:kotlin/jvm/internal/Intrinsics signature:checkNotNull(Ljava/lang/Object;Ljava/lang/String;)V"
+    ),
+    new MatchEngine(
+      "exprent type:invocation invclass:kotlin/jvm/internal/Intrinsics signature:checkNotNull(Ljava/lang/Object;)V"
+    )
+  };
+
+  // Intrinsics.areEqual($lhs$, $rhs$)
+  private static final MatchEngine EQUAL_INTRINSIC = new MatchEngine(
+    "exprent type:invocation invclass:kotlin/jvm/internal/Intrinsics name:areEqual parameter:0:$lhs$ parameter:1:$rhs$"
+  );
+
+  // ($x$ != null) ? $x$ : $y$
+  private static final MatchEngine TERNARY_NULL_CHECK = new MatchEngine(
+    "exprent type:function functype:ternary",
+    " exprent position:1 ret:$x1$",
+    " exprent position:2 ret:$y$",
+    " exprent position:0 type:function functype:neq",
+    "  exprent ret:$x$",
+    "  exprent type:constant consttype:null"
+  );
+
+  private static class ResugarRes {
+    public final Exprent expr;
+    public final boolean remove;
+
+    public ResugarRes(Exprent expr) {
+      this(expr, false);
+    }
+
+    public ResugarRes(Exprent expr, boolean remove) {
+      this.expr = expr;
+      this.remove = remove;
+    }
+  }
+
+  private static ResugarRes resugarExpr(Exprent ex) {
+    for (MatchEngine engine : NONNULL_INTRINSICS) {
+      if (engine.match(ex)) {
+        return new ResugarRes(null, true);
+      }
+    }
+
+    if (EQUAL_INTRINSIC.match(ex)) {
```

Comment:
are these not thread save?

Reply:
Good point- however I'm not particularly sure how to make this threadsafe apart from simply using a threadlocal